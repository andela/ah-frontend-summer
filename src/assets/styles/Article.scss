.ArticleBody {
    padding: 30px;
    font-size: 1.2rem;
}

.SubmainContainer {
    display: grid;
    grid-template-columns: 1fr 1fr;
    padding: 20px;
    align-content: center;
    justify-content: center
}

.actionButtons {
    padding: 10px;
}

.LikeDislikeButton {
    border: none;
    cursor: pointer;
    transition: .3s all;
    margin-left: 15px;
    outline: none;
    background: transparent;
}

.LikeDislikeButton:active{
    transform: translateY(2px);
}
