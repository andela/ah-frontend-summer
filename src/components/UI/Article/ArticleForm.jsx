import React from 'react';
import {
    Container,
    Header,
    Segment,
    Form,
    Button,
    Label
} from 'semantic-ui-react';
import CKEditor from '@ckeditor/ckeditor5-react';
import ClassicEditor from '@ckeditor/ckeditor5-build-classic';
import { WithContext as ReactTags } from 'react-tag-input';

import '../../../assets/styles/CreateArticleForm.scss';
import '../../../assets/styles/SignupForm.scss';
import '../../../assets/styles/TagsInput.scss';

const editorConfig = {
    toolbar: [
        'bold',
        'italic',
        '|',
        'bulletedList',
        'numberedList',
        'blockQuote',
        '|',
        'undo',
        'redo'
    ],
    placeholder: 'Create your Article...'
};

const KeyCodes = {
    comma: 188,
    enter: 13,
};
const delimiters = [KeyCodes.comma, KeyCodes.enter];

const articleForm = (props) => {
    const {
        formTitle,
        title,
        description,
        inputChanged,
        onImageUpload,
        submitData,
        loading,
        disabled,
        buttonAction,
        suggestions,
        tags,
        handleDelete,
        handleAddition,
        handleDrag,
        handleTagClick,
        createMode
    } = props;
    return (
        <div className="CreateArticleForm">
            <Container>
                <Segment padded="very">
                    <Header as="h2" textAlign="center" color="teal">
                        {formTitle}
                    </Header>
                    <Form padded="very">
                        <div className="spaceComponents">
                            <Label as="a" color="teal" ribbon>
                                Title
                            </Label>
                            <Form.Input
                                size="large"
                                name="title"
                                fluid
                                required
                                onChange={inputChanged}
                                value={title}
                            />
                            <Label as="a" color="teal" ribbon>
                                Description
                            </Label>
                            <Form.Input
                                size="large"
                                name="description"
                                fluid
                                required
                                onChange={inputChanged}
                                value={description}
                            />
                            <Label as="a" color="teal" ribbon>
                                    Body
                            </Label>
                            <CKEditor
                                editor={ClassicEditor}
                                config={editorConfig}
                                data={props.body}
                                onChange={props.bodyChanged}
                                name="boddyEditor"
                            />
                            <Label as="a" color="teal" ribbon>
                                    Image
                            </Label>
                            <div className="ImageInputField">
                                <input
                                    type="file"
                                    id="single"
                                    onChange={onImageUpload}
                                />
                            </div>
                            { createMode ? (
                                <ReactTags
                                    tags={tags}
                                    suggestions={suggestions}
                                    delimiters={delimiters}
                                    handleDelete={handleDelete}
                                    handleAddition={handleAddition}
                                    handleDrag={handleDrag}
                                    handleTagClick={handleTagClick}
                                />
                            ) : null
                            }
                            <Button
                                fluid
                                basic
                                color="teal"
                                onClick={submitData}
                                loading={loading}
                                disabled={disabled}
                                size="huge">
                                {buttonAction}
                            </Button>
                        </div>
                    </Form>
                        
                </Segment>
            </Container>
        </div>
    );
};

export default articleForm;
