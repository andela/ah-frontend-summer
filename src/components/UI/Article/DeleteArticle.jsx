import React from 'react';
import { Button, Container } from 'semantic-ui-react'

const deleteArticleButtons = (props) => {
    const {deleteArticle, dismissModal } = props;
    return (
        <Container textAlign="center">
            <div>
                <p>Are you sure you want to delete this article?</p>
                <Button
                    color='red'
                    onClick={deleteArticle}
                    >Yes</Button> 
                <Button
                    color='blue'
                    onClick={dismissModal}
                    >No</Button>
            </div>
        </Container>
    );
};

export default deleteArticleButtons
