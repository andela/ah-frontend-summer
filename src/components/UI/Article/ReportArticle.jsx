import React from 'react';
import { connect } from 'react-redux';
import {
    Container,
    Segment,
    Form,
    Button,
    Radio,
    GridColumn,
    GridRow,
    Grid
} from 'semantic-ui-react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';

export const ReportArticle = props => {
    const { onSubmitHandler, onChangeHandler, slug, reason, loading } = props;
    const content = (
        <GridColumn width={8}>
            <Form onSubmit={onSubmitHandler}>
                <h3>Report Article</h3>
                <Form.Field>
                    <Radio
                        label='Spam'
                        name='spam'
                        value='spam'
                        checked={reason === 'spam'}
                        onChange={onChangeHandler}
                    />
                </Form.Field>
                <Form.Field>
                    <Radio
                        label='Harassment'
                        name='harassment'
                        value='harassment'
                        checked={reason === 'harassment'}
                        onChange={onChangeHandler}
                    />
                </Form.Field>
                <Form.Field>
                    <Radio
                        label='Rules Violation'
                        name='rules violation'
                        value='rules violation'
                        checked={reason === 'rules violation'}
                        onChange={onChangeHandler}
                    />
                </Form.Field>
                <Button basic color='teal' type='submit' loading={loading}>
                    Report
                </Button>
                <Link to={`/articles/${slug}`}>
                    <Button basic>Cancel</Button>
                </Link>
            </Form>
        </GridColumn>
    );

    return (
        <Container>
            <Grid>
                <GridRow>
                    <GridColumn width={4}></GridColumn>
                    <GridColumn width={8}>
                        <Segment padded='very' color='teal'>
                            <Grid>
                                <GridRow>
                                    <GridColumn width={4}></GridColumn>
                                    {content}
                                    <GridColumn width={4}></GridColumn>
                                </GridRow>
                            </Grid>
                        </Segment>
                    </GridColumn>
                    <GridColumn width={4}></GridColumn>
                </GridRow>
            </Grid>
        </Container>
    );
};

ReportArticle.propTypes = {
    onSubmitHandler: PropTypes.func.isRequired,
    onChangeHandler: PropTypes.func.isRequired,
    slug: PropTypes.string.isRequired,
    reason: PropTypes.string.isRequired
};

const mapStateToProps = state => {
    return {
        loading: state.reportArticle.loading
    };
};

export default connect(mapStateToProps)(ReportArticle);
