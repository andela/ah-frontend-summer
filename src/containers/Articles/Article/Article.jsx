import React, { Component } from "react";
import { connect } from "react-redux";
import { Redirect } from "react-router-dom"
import { toast } from "react-toastify";

import * as actions from "../../../store/actions/index";
import ArticleDetails from "../../../components/UI/Article/Article";
import DeleteArticleButtons from "../../../components/UI/Article/DeleteArticle";
import Loader from "../../../components/UI/Loader";
import {dangerToast, infoToast } from "../../../components/Toast/toast";
import Modal from "../../../components/UI/Modal/Modal";

export class Article extends Component {

    state = {
        showModal: false
    }

    componentWillMount(){
        this.props.onGetArticle(this.props.match.params.slug);
    }

    deleteArticleHandeler = () => {
        toast.dismiss();
        this.props.onDeleteArticle(this.props.match.params.slug);
    }

    showModalHandeler = () => {
        this.setState({ showModal: true });
    }

    closeModalHandler = () => {
        this.setState({ showModal: false });
    }

    render() {
        this.renderArticle = ( <Loader /> );
        this.created_at = null;
        this.isAuthor = false;
        this.isAuthenticated = localStorage.getItem("token") !== null;
        this.username = localStorage.getItem("username");
        const { article, loading, error } = this.props;
        this.confirmButtons = null;
        if (this.state.showModal) {
            this.confirmButtons = (
                <DeleteArticleButtons
                    dismissModal={this.closeModalHandler}
                    deleteArticle={this.deleteArticleHandeler}/>
            );
        };

        if ( article && article.article === "Article has been deleted") {
            infoToast(article.article);
            this.renderArticle = (<Redirect to="/"/>);
        } else if (article) {
            if (this.username) {
                this.isAuthor = ( this.username ===
                    article.author.username
                    && this.isAuthenticated)
            };
            this.created_at = article.created_at.substr(0,
                article.created_at.indexOf("T"));
            this.renderArticle = (
                <ArticleDetails
                    title={article.title}
                    author={article.author.username}
                    body={article.body}
                    readTime={article.read_time}
                    deleteArticle={this.showModalHandeler}
                    slug={this.props.match.params.slug}
                    loading={loading}
                    dateCreated={this.created_at}
                    isAuthor={this.isAuthor}
                    image={article.image}
                    />);
        } else if ( error ) {
            dangerToast(error.articles.errors);
            this.renderArticle = (<Redirect to="/"/>);
        };

        return (
            <div>
                <Modal
                show={this.state.showModal}
                modalClosed={this.closeModalHandler}>{this.confirmButtons}</Modal>
                {this.renderArticle}
            </div>
        );
    }
}

export const mapStateToProps = (state) => {
    return {
        article: state.article.article,
        loading: state.article.loading,
        error: state.article.error
    };
};

export const mapDispatchToProps = (dispatch) => {
    return {
        onGetArticle: (slug) => dispatch(actions.getArticle(slug)),
        onDeleteArticle: (slug) => dispatch(actions.deleteArticle(slug))
    };
};

export default connect(mapStateToProps, mapDispatchToProps)(Article);
