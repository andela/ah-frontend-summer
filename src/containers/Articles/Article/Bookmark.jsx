import React, {Component} from 'react';
import {connect} from "react-redux";
import PropTypes from "prop-types";
import {withRouter} from "react-router-dom";
import Bookmark from "../../../components/article/Bookmark";
import {
    bookmarkAction,
    undoBookmarkAction
} from "../../../store/actions/async/ArticleActions";
import {isLoggedIn} from "../../../store/actions/async";

export class BookmarkContainer extends Component{
    bookmarkArticle = () => {
        const {history} = this.props;
        const loggedIn = isLoggedIn(localStorage);
        if (!loggedIn){
            history.push('/login');
        }
        else{
            this.triggerBookmark();
        }
    };

    triggerBookmark = () => {
        const {bookmark, undoBookmark, isBookmarked, slug, loading} = this.props;

        if (!loading){
            if (isBookmarked){
                undoBookmark(slug);
            }
            else {
                bookmark(slug);
            }
        }
    };

    render() {
        const {loading, isBookmarked} = this.props;
        return (
            <Bookmark
                loading={loading}
                bookmark={() => {this.bookmarkArticle();}}
                isBookmarked={isBookmarked} />
        );
    }
}

export const mapStateToProps = (state) => {
    const {loading, status} = state.progress.bookmark;
    const {favorited: isBookmarked} = state.article.article;
    return {loading, status, isBookmarked: isBookmarked};
};

export const mapDispatchToProps = (dispatch) => {
    return {
        bookmark: (slug) => {
            dispatch(bookmarkAction(slug));
        },
        undoBookmark: (slug) => {
            dispatch(undoBookmarkAction(slug));
        }
    };
};

export default withRouter(connect(mapStateToProps, mapDispatchToProps)(BookmarkContainer));

BookmarkContainer.propTypes = {
    bookmark: PropTypes.func.isRequired,
    history: PropTypes.object.isRequired,
    isBookmarked: PropTypes.bool.isRequired,
    loading: PropTypes.bool.isRequired,
    slug: PropTypes.string.isRequired,
    undoBookmark: PropTypes.func.isRequired
};
