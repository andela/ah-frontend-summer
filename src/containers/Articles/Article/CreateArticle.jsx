import React, { Component } from "react";
import { connect } from "react-redux";
import { EditorState } from "draft-js";
import { Redirect } from "react-router-dom";
import PropTypes from "prop-types";

import * as actions from "../../../store/actions/index";
import CreateArticleForm from "../../../components/UI/Article/ArticleForm";
import { successToast } from "../../../components/Toast/toast";
import sampleTags from '../../../assets/utils/sampleTags';

const suggestions = sampleTags.map((tag) => {
    return {
        id: tag,
        text: tag
    };
});
export class CreateArticle extends Component {

    state = {
        editorState: EditorState.createEmpty(),
        articleData: {
            title: "",
            description: "",
            body: "",
            image: null
        },
        error: false,
        titleTouched: false,
        descriptionTouched: false,
        bodyTouched: false,
        disabled: false,
        suggestions: suggestions,
        tags: []
    };

    onChangeHandler = (editorState) => {
        this.setState({editorState});
    };

    bodyChangedHandler=( event, editor ) => {
        const data = editor.getData();
        const articleData = {...this.state.articleData};
        articleData.body = data;
        const { body, title, description } = articleData;
        let disableBtn = true;
        if ( !body || body.length < 5 || !title || !description) {
            disableBtn = false;
        }

        this.setState({
            articleData,
            disabled: disableBtn
        });
    }

    inputHandler = (e) => {
        const articleData = {...this.state.articleData};
        let titleTouched = false;
        let descriptionTouched = false;
        let bodyTouched = false;
        let disableBtn = true;
        const { name, value, files } = e.target;

        if (name === "title") {
            articleData.title = value;
            titleTouched = false;
        } else if (name === "description") {
            articleData.description = value;
            descriptionTouched = false;
        }
        const { body, title, description } = articleData;

        if (title) {
            disableBtn = title.length > 3 && disableBtn;
        }
        if (description) {
            disableBtn = description.length > 3 && disableBtn;
        }
        if ( !title || !description || !body) {
            disableBtn = false;
        }
        if (files) {
            articleData.image = files[0];
        }

        this.setState ({
            articleData,
            titleTouched,
            descriptionTouched,
            bodyTouched,
            disabled: disableBtn
        });
        e.preventDefault();
    }

    createArticleSubmitHandler = () => {
        let disableBtnAgain = false;
        this.setState({ disabled: disableBtnAgain });
        const {
            image,
            title,
            body,
            description
        } = this.state.articleData;
        const { tags } = this.state;
        const dataToFormData = new FormData();
        if (image !== null) {
            dataToFormData.append("image", image, image.name);
        }
        if (tags !== undefined || tags.length !== 0) {
            tags.map(tag => {
                dataToFormData.append("tag_list", tag.text);
                return tag.text;
            });
        }
        dataToFormData.append("title", title);
        dataToFormData.append("body", body);
        dataToFormData.append("description", description);

        this.props.onCreateArticle(dataToFormData);
    }

    handleDelete = i => {
        const { tags } = this.state;
        this.setState({
            tags: tags.filter((tag, index) => index !== i),
        });
    }
    
    handleAddition = tag => {
        this.setState(state => ({ tags: [...state.tags, tag] }));
    }
    
    handleDrag = (tag, currPos, newPos) => {
        const { tags } = this.state;
        const stateTags = [...tags];
        const newTags = stateTags.slice();
    
        newTags.splice(currPos, 1);
        newTags.splice(newPos, 0, tag);
    
        // re-render
        this.setState({ tags: newTags });
    }

    render() {
        this.isAuthenticated = localStorage.getItem("token") !== null;
        const { loading, message } = this.props;
        const { disabled, editorState, suggestions, tags, articleData } = this.state;
        const { title, description, body } = articleData;

        this.createArticle = (
            <div>
                <CreateArticleForm
                    editorState={editorState}
                    onChange={this.onChangeHandler}
                    inputChanged={this.inputHandler}
                    submitData={this.createArticleSubmitHandler}
                    buttonAction="Create Article"
                    formTitle="Create an article with Authors Haven"
                    title={title}
                    description={description}
                    body={body}
                    loading={loading}
                    disabled={!disabled}
                    bodyChanged={this.bodyChangedHandler}
                    onImageUpload={this.inputHandler}
                    suggestions={suggestions}
                    tags={tags}
                    handleDelete={this.handleDelete}
                    handleAddition={this.handleAddition}
                    handleDrag={this.handleDrag}
                    handleTagClick={this.handleTagClick}
                    createMode
                />
            </div>
        );
        if (
            message && message ===
            "Article created successfully") {
            successToast(message);
            this.createArticle = (<Redirect to="/" />);
        } else if (!this.isAuthenticated) {
            this.createArticle = (<Redirect to="/login" />);
        }

        return (
            <div>
                {this.createArticle}
            </div>
        );
    }
}

export const mapStateToProps = state => {
    return {
        loading: state.article.loading,
        article: state.article.article,
        message: state.article.message
    };
};

export const mapDispatchToProps = dispatch => {
    return {
        onCreateArticle: data => dispatch(actions.createArticle(data))
    };
};

CreateArticle.propTypes = {
    loading: PropTypes.string.isRequired,
    message: PropTypes.string.isRequired
};

export default connect(mapStateToProps, mapDispatchToProps)(CreateArticle);
