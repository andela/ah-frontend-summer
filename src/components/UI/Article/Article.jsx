import React from 'react';
import { Container, Button, Header, Image, Icon } from 'semantic-ui-react';
import { Link } from 'react-router-dom';
import renderHTML from 'react-render-html';

import '../../../assets/styles/Article.scss';

const article = (props) => {
    const {
        title,
        author,
        dateCreated,
        readTime,
        isAuthor,
        deleteArticle,
        slug,
        loading,
        image,
        body,
        facebook,
        twitter,
        email
    } = props;
    const renderedBody = renderHTML(body)
    const defautImage = 'https://images.pexels.com/photos/1591057/pexels-photo-1591057.jpeg?cs=srgb&dl=buttons-characters-concept-1591057.jpg&fm=jpg';
    const isLoggedIn = localStorage.getItem('token');

    return (
        <div>
            <Container>
                <Header as='h1' padded="very">
                    {title}
                </Header>
            </Container>
            <Container>
                {author}
                <Container>
                    {dateCreated}
                    <p>{readTime}</p>
                </Container>
            </Container>
            {isAuthor ? (
                <div className="actionButtons">
                    <Container textAlign='right'>
                        <Button
                            color="red"
                            basic
                            onClick={deleteArticle}>
                            Delete
                        </Button>
                        <Link className="article-card-title" to={`/articles/update/${slug}`}>
                            <Button color="teal" basic loading={loading}>Update</Button>
                        </Link>
                    </Container>
                </div>
            ) : null
            }
            <Container textAlign='justified'>
                <Image src={image ? image : defautImage} />
                <div className="ArticleBody">
                    {renderedBody}
                </div>
            </Container>
            {isLoggedIn ? (
                <Container textAlign="center">
                    <a href={facebook} className="facebook"><Icon name="facebook f" /></a>
                    <a href={twitter} className="twitter"><Icon name="twitter" /></a>
                    <a href={email} className="email"><Icon name="mail" /></a>
                </Container>
            ) : ""}
        </div>
    );
};

export default article;
