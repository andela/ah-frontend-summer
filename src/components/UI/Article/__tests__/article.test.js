import React from 'react';
import { shallow } from 'enzyme';

import Article from '../Article';

describe('<Article />', () => {
    it('Article will render without crashing', () => {
        const props = {
            body: "body",
            tags:[]
        };
        let wrapper = shallow(<Article {...props}/>);
        expect(wrapper).toMatchSnapshot();
    });
});
