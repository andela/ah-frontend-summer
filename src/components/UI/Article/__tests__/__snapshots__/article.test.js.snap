// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Article /> Article will render without crashing 1`] = `
<div>
  <Container>
    <Header
      as="h1"
      padded="very"
    />
  </Container>
  <Container>
    <Container>
      <p />
    </Container>
  </Container>
  <Container
    textAlign="justified"
  >
    <Image
      as="img"
      src="https://images.pexels.com/photos/1591057/pexels-photo-1591057.jpeg?cs=srgb&dl=buttons-characters-concept-1591057.jpg&fm=jpg"
      ui={true}
    />
    <div
      className="ArticleBody"
    >
      body
    </div>
  </Container>
</div>
`;
